### PR TITLE
Add the ClassLoaderTypeSolver

### PR DESCRIPTION
**Pull request by ftomassetti**
As suggested in #1745 we add the possibility to use custom classloaders wrapped as typesolvers

### CODE REVIEW

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+import com.github.javaparser.resolution.declarations.ResolvedReferenceTypeDeclaration;
+import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
+import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
+import com.github.javaparser.symbolsolver.reflectionmodel.ReflectionFactory;
+
+import java.util.Optional;
+
+/**
+ * This TypeSolver wraps a ClassLoader. It can solve all types that the given ClassLoader can solve.
```
Comment (matozoid) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java]: .. that the ClassLoader can *load*.

**Comment by matozoid** [comment]
+1

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    public void setParent(TypeSolver parent) {
+        this.parent = parent;
+    }
+
+    protected boolean filterName(String name) {
+        return true;
+    }
+
+    @Override
+    public SymbolReference<ResolvedReferenceTypeDeclaration> tryToSolveType(String name) {
```
Comment (jGauravGupta) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java:57]: If JavaParser API is invoked to resolve the class (for e.g my.company.SampleClass) then this method also receive the input call where **name** param contains value "java.lang.my.company.SampleClass".

Comment (jGauravGupta) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java:57]: Perhaps performance improvement, if this type of redundant call can be fixed.

Comment (ftomassetti) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java:57]: I am not sure I understand

Comment (jGauravGupta) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java:57]: I will create separate ticket with more info as this issue is applicable to ReflectionTypeResolver too, btw this PR looks good to me for Classloader type resolver support.
-Thanks

Comment (ftomassetti) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/ClassLoaderTypeSolver.java:57]: Ok, then I will merge this one and work on the issues you will submit later. Thanks!

**Pull request merged**